apis/mutations/unversioned/modifyset_types.go

```diff
@@ -77,37 +80,92 @@ const (
 // Values describes the values provided to the operation.
 // +kubebuilder:object:generate=false
 type Values struct {
+	// FromList is a list of values to be merged or pruned.
 	FromList []interface{} `json:"fromList,omitempty"`
+
+	// ExternalData describes the external data source to be used for mutation.
+	ExternalData *ExternalData `json:"externalData,omitempty"`
 }
 
-func (in *Values) DeepCopy() *Values {
-	if in == nil {
+func (v *Values) DeepCopy() *Values {
+	if v == nil {
 		return nil
 	}
 
 	var fromList []interface{}
-	if in.FromList != nil {
-		fromList = make([]interface{}, len(in.FromList))
+	if v.FromList != nil {
+		fromList = make([]interface{}, len(v.FromList))
 		for i := range fromList {
-			fromList[i] = runtime.DeepCopyJSONValue(in.FromList[i])
+			fromList[i] = runtime.DeepCopyJSONValue(v.FromList[i])
 		}
 	}
 
+	var externalData *ExternalData
+	if v.ExternalData != nil {
+		externalData = v.ExternalData.DeepCopy()
+	}
+
 	return &Values{
-		FromList: fromList,
+		FromList:     fromList,
+		ExternalData: externalData,
 	}
 }
 
-func (in *Values) DeepCopyInto(out *Values) {
-	*in = *out
+func (v *Values) DeepCopyInto(out *Values) {
+	*v = *out
 
-	if in.FromList != nil {
-		fromList := make([]interface{}, len(in.FromList))
+	if v.FromList != nil {
+		fromList := make([]interface{}, len(v.FromList))
 		for i := range fromList {
-			fromList[i] = runtime.DeepCopyJSONValue(in.FromList[i])
+			fromList[i] = runtime.DeepCopyJSONValue(v.FromList[i])
 		}
 		out.FromList = fromList
 	}
+
+	if v.ExternalData != nil {
+		out.ExternalData = v.ExternalData.DeepCopy()
+	}
+}
+
+// Validate validates the values.
+func (v *Values) Validate() error {
+	if v.FromList == nil && v.ExternalData == nil {
+		return fmt.Errorf("values must set one of `fromList` or `externalData`")
+	}
+
+	if v.FromList != nil && v.ExternalData != nil {
+		return fmt.Errorf("values must only set one of `fromList` or `externalData`")
+	}
+
+	if v.ExternalData != nil {
+		return v.ExternalData.Validate()
+	}
+
+	return nil
+}
+
+// Get returns the list of values to be merged or pruned.
+func (v *Values) Get(param *types.GetValueParameter) ([]interface{}, error) {
```

Comment:
How does external data work with modify set? What is the value that gets passed to the external data provider?

Reply:
Based on the comment below, if we only support string object, we won't be able to support modify set with external data.